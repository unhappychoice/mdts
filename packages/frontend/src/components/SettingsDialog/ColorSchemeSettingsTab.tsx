import { Box, MenuItem, Select, ToggleButton, ToggleButtonGroup, Typography } from '@mui/material';
import { useTheme } from '@mui/material/styles';
import React, { useCallback } from 'react';
import {
  SYNTAX_HIGHLIGHTER_THEMES
} from '../../constants';
import { themes } from '../../styles/themes';

interface ColorSchemeSettingsTabProps {
  darkMode: 'dark' | 'light' | 'auto';
  theme: string;
  syntaxHighlighterTheme: string;
  handleToggleDarkMode: (mode: 'dark' | 'light' | 'auto') => void;
  setSyntaxHighlighterTheme: (theme: string) => void;
  handleToggleTheme: (theme: string) => void;
}

const ColorSchemeSettingsTab: React.FC<ColorSchemeSettingsTabProps> = ({
  darkMode,
  syntaxHighlighterTheme,
  theme,
  handleToggleDarkMode,
  setSyntaxHighlighterTheme,
  handleToggleTheme,
}) => {
  const muiTheme = useTheme();

  const handleDarkModeChange = useCallback((_, newMode) => {
    if (newMode)
      handleToggleDarkMode(newMode);
  }, [handleToggleDarkMode]);

  const handleThemChange = useCallback((e) => {
    setSyntaxHighlighterTheme(e.target.value as string);
  }, [setSyntaxHighlighterTheme]);

  const handleThemeChange = useCallback((e) => {
    handleToggleTheme(e.target.value as string);
  }, [handleToggleTheme]);

  return (
    <Box sx={{ p: 2 }}>
      <Typography variant="subtitle2" sx={{ mb: 2 }}>Dark Mode</Typography>
      <ToggleButtonGroup
        value={darkMode}
        exclusive
        onChange={handleDarkModeChange}
        aria-label="dark mode"
        fullWidth
        size="small"
        sx={{ mb: 2 }}
      >
        <ToggleButton
          value="dark"
          aria-label="dark"
          sx={{
            flexGrow: 1,
            '&.Mui-selected': {
              backgroundColor: muiTheme.palette.primary.main,
              color: muiTheme.palette.primary.contrastText,
              '&:hover': {
                backgroundColor: muiTheme.palette.primary.dark,
              },
            },
          }}
        >
          Dark
        </ToggleButton>
        <ToggleButton
          value="light"
          aria-label="light"
          sx={{
            flexGrow: 1,
            '&.Mui-selected': {
              backgroundColor: muiTheme.palette.primary.main,
              color: muiTheme.palette.primary.contrastText,
              '&:hover': {
                backgroundColor: muiTheme.palette.primary.dark,
              },
            },
          }}
        >
          Light
        </ToggleButton>
        <ToggleButton
          value="auto"
          aria-label="auto"
          sx={{
            flexGrow: 1,
            '&.Mui-selected': {
              backgroundColor: muiTheme.palette.primary.main,
              color: muiTheme.palette.primary.contrastText,
              '&:hover': {
                backgroundColor: muiTheme.palette.primary.dark,
              },
            },
          }}
        >
          Auto (System)
        </ToggleButton>
      </ToggleButtonGroup>

      <Typography variant="subtitle2" sx={{ mb: 2 }}>Theme</Typography>
      <Select
        value={theme}
        onChange={handleThemeChange}
        fullWidth
        variant="outlined"
        size="small"
        sx={{ mb: 2 }}
      >
        {themes.map(t => {
          const palette = (darkMode === 'dark' || (darkMode === 'auto' && muiTheme.palette.mode === 'dark')) ? t.dark : t.light;
          const background = `linear-gradient(45deg, ${palette.background} 0%, ${palette.paper} 100%)`;

          return (
            <MenuItem key={t.id} value={t.id}>
              <Box sx={{ display: 'flex', alignItems: 'center', width: '100%' }}>
                <Box sx={{ display: 'flex', alignItems: 'center', mr: 1.5, gap: '4px' }}>
                  <Box
                    sx={{
                      width: 20,
                      height: 20,
                      borderRadius: '4px',
                      background: background,
                      border: `1px solid ${muiTheme.palette.divider}`,
                    }}
                  />
                </Box>
                <Typography variant="body2">{t.name}</Typography>
              </Box>
            </MenuItem>
          );
        })}
      </Select>

      <Typography variant="subtitle2" sx={{ my: 2 }} id="syntax-highlighter-theme-label">Syntax Highlighter Theme</Typography>
      <Select
        labelId="syntax-highlighter-theme-label"
        id="syntaxHighlighterTheme"
        value={syntaxHighlighterTheme || 'auto'}
        onChange={handleThemChange}
        fullWidth
        variant="outlined"
        size="small"
        sx={{ mb: 1 }}
        data-testid="syntax-highlighter-theme-select"
      >
        <MenuItem value="auto">Auto</MenuItem>
        {SYNTAX_HIGHLIGHTER_THEMES.map(theme => {
          const backgroundColor = theme.color || muiTheme.palette.background.default; // Fallback to default background

          return (
            <MenuItem key={theme.value} value={theme.value}>
              <Box sx={{ display: 'flex', alignItems: 'center', width: '100%' }}>
                <Box sx={{ display: 'flex', alignItems: 'center', mr: 1.5, gap: '4px' }}>
                  <Box
                    sx={{
                      width: 20,
                      height: 20,
                      borderRadius: '4px',
                      background: backgroundColor,
                      border: `1px solid ${muiTheme.palette.divider}`,
                    }}
                  />
                </Box>
                <Typography variant="body2">{theme.name}</Typography>
              </Box>
            </MenuItem>
          );
        })}
      </Select>
    </Box>
  );
};

export default ColorSchemeSettingsTab;
