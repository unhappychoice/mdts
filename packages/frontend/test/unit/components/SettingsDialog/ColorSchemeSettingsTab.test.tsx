import { render } from '@testing-library/react';
import React from 'react';
import ColorSchemeSettingsTab from '../../../../src/components/SettingsDialog/ColorSchemeSettingsTab';

describe('ColorSchemeSettingsTab', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<ColorSchemeSettingsTab />);
    expect(asFragment()).toMatchSnapshot();
  });
});
