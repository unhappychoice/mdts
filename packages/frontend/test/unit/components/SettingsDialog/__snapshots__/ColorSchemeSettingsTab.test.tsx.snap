// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ColorSchemeSettingsTab should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-19midj6"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 css-n2une3-MuiTypography-root"
    >
      Dark Mode
    </h6>
    <div
      aria-label="dark mode"
      class="MuiToggleButtonGroup-root MuiToggleButtonGroup-horizontal MuiToggleButtonGroup-fullWidth css-1jjbo4h-MuiToggleButtonGroup-root"
      role="group"
    >
      <button
        aria-label="dark"
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-firstButton css-zhjb61-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="dark"
      >
        Dark
      </button>
      <button
        aria-label="light"
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-zhjb61-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="light"
      >
        Light
      </button>
      <button
        aria-label="auto"
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-lastButton css-zhjb61-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="auto"
      >
        Auto (System)
      </button>
    </div>
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 css-n2une3-MuiTypography-root"
    >
      Theme
    </h6>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-sizeSmall MuiSelect-root css-1v0ium5-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-19marfc-MuiNativeSelect-root-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        role="combobox"
        tabindex="0"
      >
        <span
          aria-hidden="true"
          class="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
        tabindex="-1"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-1rfqz7c-MuiSvgIcon-root-MuiNativeSelect-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1nf2c5d-MuiNotchedOutlined-root"
        >
          <span
            aria-hidden="true"
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 css-14l8bx7-MuiTypography-root"
      id="syntax-highlighter-theme-label"
    >
      Syntax Highlighter Theme
    </h6>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-sizeSmall MuiSelect-root css-134kgh-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      data-testid="syntax-highlighter-theme-select"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="syntax-highlighter-theme-label syntaxHighlighterTheme"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-19marfc-MuiNativeSelect-root-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="syntaxHighlighterTheme"
        role="combobox"
        tabindex="0"
      >
        Auto
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
        tabindex="-1"
        value="auto"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-1rfqz7c-MuiSvgIcon-root-MuiNativeSelect-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1nf2c5d-MuiNotchedOutlined-root"
        >
          <span
            aria-hidden="true"
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
